ledger/simulation/resources.go

```diff
@@ -0,0 +1,584 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package simulation
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// ResourceAssignment calculates the additional resources that a transaction or group could use, and
+// it tracks any referenced unnamed resources that fit within those limits.
+type ResourceAssignment struct {
+	Accounts    map[basics.Address]struct{}
+	MaxAccounts int
+
+	Assets    map[basics.AssetIndex]struct{}
+	MaxAssets int
+
+	Apps    map[basics.AppIndex]struct{}
+	MaxApps int
```

Comment:
just a nit, but I was wondering as we utilizing generics across the repo, maybe we should define a set of abstract data structure from generics, for example `set` here: https://bitfieldconsulting.com/golang/generic-set

it would be totally fine if we don't do it in this PR, but just a thing to think about

Reply:
Probably we should begin looking around to see what popular libraries are out there. It's probably too early to pick a winner, but soon we should have access to Map/Filter and so on as well. (Jason actually wrote Map(), as convertSlice() here. Someday, we'd just call into the library we settle on.)